### PR TITLE
Custom Languages Part 1

### PR DESCRIPTION
**Pull request by jhwgh1968**
- [x] Everything I'm contributing is my own work, used with permission (and attributed as necessary), or is very clearly [fair use](https://en.wikipedia.org/wiki/Fair_use)
- [x] I agree to put my contribution into the Public Domain (or CC-0, if that doesn't apply where you are)

---

This is a massive PR, I know. Hopefully it takes everything into account we discussed in other channels.

In brief, the flow is this:
1. Start with easy to understand: a postfix calculator
2. Realize it's sort of designed like a compiler if you squint
3. Make it more formally like a compiler, splitting it into in phases to do the same trivial work
4. Add a JIT library and explain its IR  **(current PR is up to here)**
6. Integrate that JIT library into the compiler we have
7. Using the library, add features of a real language: variables, flow control, subroutines, etc
8. Write a noteworthy program in the language, currently thinking AES encryption on a block of data
9. Benchmark it compared to a pure Python version and be amazed
10. Profit!

Suffice it to say, I chose this approach to embedding and formatting because otherwise I could not keep the Python reference blocks actually working together as I work on them. I'd be happy to take any organizational suggestions you have.

Also, there is one lazy ASCII art diagram in a `text` block. If you have a better approach for such "quick and dirty" diagrams, please let me know that too.

### CODE REVIEW

**Comment by jhwgh1968** [comment]
I have figured out a way to integrate graphviz diagrams, but I've put it on a separate branch. I'll let you review this first.

**Comment by VegaDeftwing** [comment]
I'm currently focusing on finishing up the AI image generation blog post (Which I would greatly appreciate feedback on! - it's in the AIBlogPost branch) before I get to this, as that space is changing enough that what I have to say is time sensitive both for relevance and because the facts of the situation may change necessitating a re-write. I will *probably* get to this this weekend though.

**Comment by VegaDeftwing** [comment]
@jhwgh1968, I've made a few changes, almost all in presentation but some very minor to content. I was able to run the first python script fine, but don't have time to setup and run the others quite yet. So far, it's been a delight to read and I'm learning a lot. Thank you so much for all the hard work you're putting into this!!

File: content/Engineering/Programming/Compilers/generation.md [hunk 1]
```diff
@@ -23,20 +29,24 @@ end while
...
- |   /                                                              \
- v  v         true                                                   \
-[ n < 10 ] ----> [ x = read integer from file; array[n] = x; n = n + 1]
-  \
-   \  false
-    ----------> [ print array[0] ] --> end
-
-```
+This would be broken into 3 blocks. Note, all the logic in the loop can be thought of as a single block, as the operations are consecutive without any conditional logic (as long as me assume "read integer from file" can't fail)
+
```
Comment (jhwgh1968) [inline content/Engineering/Programming/Compilers/generation.md:33]: A small nit: the "block" in this case is the entire contents of the loop body.

I think your diagram is much better, but perhaps call it "loop block"? Or maybe have "block" shading around the other pieces to distinguish it from a statement?

**Comment by jhwgh1968** [comment]
It is certainly looking much better, especially that diagram compared to my manual graphviz hack. I did see Mermaid, but Hugo versioning and package repos made it seem unsupported...

I'm glad you're learning and having fun! I also hope that my creating a Python module for my own testing is helpful for your testing as well. Making sure the code examples still work is really important as I continue forward.

File: content/Engineering/Programming/Compilers/generation.md (hunk 1 above)
Comment (jhwgh1968) [reply content/Engineering/Programming/Compilers/generation.md:33]: Ah, now I see it in the text above. That's good enough.

**Comment by jhwgh1968** [comment]
Hmm, I may have spoken too soon.

I intentionally did not set `language=python` on the blocks of included code because they broke. I thought your other changes had fixed that, but it seems not:

Screenshot_20221204_124231

**Comment by VegaDeftwing** [comment]
> Hmm, I may have spoken too soon.
> 
> I intentionally did not set `language=python` on the blocks of included code because they broke. I thought your other changes had fixed that, but it seems not:
> 
> Screenshot_20221204_124231

Interesting. I have no issues at all, 

image

Can you try updating your hugo install? Also, can you check on the actual, live website that the code blocks's numbers are aligned correctly for you, as it looks like your line numbers are broken locally as well, which is an issue I had a while back but have since resolved. You may also try clearing the browser cache entirely or at least ctrl+f5'ing the pages in development to see if that fixes it.

**Comment by jhwgh1968** [comment]
Wow, the Hugo version in the system pkg manager is *ancient*. Using the 10-day-old 0.107 release does indeed fix it!

With that in place, it all looks good! Let me know if you have questions/clarifications.

**Comment by VegaDeftwing** [comment]
Do you want this to be merged now or for me to wait for Pt. 2? 

I intend to get to testing all of the code very soon and will certainly do so before I click "Merge"

**Comment by VegaDeftwing** [comment]
Closes #33

**Comment by jhwgh1968** [comment]
> Do you want this to be merged now or for me to wait for Pt. 2?

I do indeed. Part 2 (and maybe 3) will be a while yet, and I broke it here because I thought it would be somewhat useful standing on its own.

> I intend to get to testing all of the code very soon and will certainly do so before I click "Merge"

Please do, and let me know if you have any issues.

> Closes # 33

At first I was about to ask, then I realized it was just the section headings, and this PR will indeed do that.

**Comment by VegaDeftwing** [comment]
This has been merged, but won't be live until the next time I publish changes to the site.

**Pull request merged**